### PR TITLE
Add gateway and allow rule schemas

### PR DESCRIPTION
**Pull request by akrousset**

### CODE REVIEW

File: apps/fz_http/priv/repo/migrations/20221207180310_create_gateways.exs [hunk 1]
```diff
@@ -4,14 +4,15 @@ defmodule FzHttp.Repo.Migrations.CreateGateways do
   def change do
     create table(:gateways, primary_key: false) do
       add(:id, :uuid, primary_key: true, default: fragment("gen_random_uuid()"))
-      add(:name, :string, null: false)
+      add(:name, :string, null: false, default: "default")
```
Comment (jamilbk) [inline apps/fz_http/priv/repo/migrations/20221207180310_create_gateways.exs]: Maybe it's better to omit the default here to always require a name.

File: apps/fz_http/priv/repo/migrations/20221207180310_create_gateways.exs [hunk 2]
```diff
@@ -4,14 +4,15 @@ defmodule FzHttp.Repo.Migrations.CreateGateways do
...
       add(:ipv4_masquerade, :boolean, null: false, default: true)
       add(:ipv6_masquerade, :boolean, null: false, default: true)
       add(:ipv4_address, :inet, null: false, default: fragment("'10.3.2.1'::inet"))
       add(:ipv6_address, :inet, null: false, default: fragment("'fd00::3:2:1'::inet"))
       add(:mtu, :integer, null: false, default: 1280)
       add(:public_key, :string, default: nil)
-      add(:registration_token, :string, default: nil)
+      add(:registration_token, :string, null: false)
+      add(:registration_token_created_at, :utc_datetime_usec, null: false)
```
Comment (jamilbk) [inline apps/fz_http/priv/repo/migrations/20221207180310_create_gateways.exs:15]:
```suggestion
      add(:registration_token_created_at, :utc_datetime_usec, null: false)
      
      create unique_index(:devices, [:name])
```

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    import FzHttp.{AllowRulesFixtures, GatewaysFixtures, UsersFixtures}
+
+    test "list_allow_rules/0 returns all allow rules" do
+      allow_rule = allow_rule()
+      assert AllowRules.list_allow_rules() == [allow_rule]
+    end
+
+    test "list_allow_rules/1 returns allow rules scoped to a gateway" do
+      gateway = gateway(%{name: "gateway"})
+      allow_rule = allow_rule(%{gateway_id: gateway.id})
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs:18]: I think you need to create another `allow_rule()` belonging to another gateway for this test to be robust.

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      assert AllowRules.list_allow_rules(gateway) == [allow_rule]
+    end
+
+    test "list_allow_rules/1 returns allow rules scoped to a user" do
+      user = user()
+      allow_rule = allow_rule(%{user_id: user.uuid})
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs:25]: I think you need to create another allow rule belonging to another user to ensure this is robust.

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      assert {:error, %Ecto.Changeset{}} = AllowRules.create_allow_rule(invalid_gateway)
+    end
+
+    test "create_allow_rule/1 with invalid destination returns an error changeset" do
+      invalid_destination = %{
+        gateway_id: gateway().id,
+        destination: nil
+      }
+
+      assert {:error, %Ecto.Changeset{}} = AllowRules.create_allow_rule(invalid_destination)
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs:72]: Probably want to use `==` to explicitly check for the error message you're expecting.

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      assert allow_rule.port_range_end == 2
+    end
+
+    test "create_allow_rule/1 with invalid gateway returns an error changeset" do
+      invalid_gateway = %{
+        gateway_id: nil,
+        destination: "10.10.10.0/24"
+      }
+
+      assert {:error, %Ecto.Changeset{}} = AllowRules.create_allow_rule(invalid_gateway)
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs:63]: Explicit error check with `==`

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    test "create_allow_rule/1 with invalid port range start returns an error changeset" do
+      invalid_port_range = %{
+        gateway_id: gateway().id,
+        destination: "10.10.10.0/24",
+        port_range_start: nil,
+        port_range_end: 2
+      }
+
+      assert {:error, %Ecto.Changeset{} = change} =
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs]: Explicit error check

File: apps/fz_http/test/fz_http/allow_rules_test.exs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    test "create_allow_rule/1 with invalid port range end returns an error changeset" do
+      invalid_port_range = %{
+        gateway_id: gateway().id,
+        destination: "10.10.10.0/24",
+        port_range_start: 1,
+        port_range_end: 65_536
+      }
+
+      assert {:error, %Ecto.Changeset{} = change} =
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/allow_rules_test.exs]: Explicit error check `==`

File: apps/fz_http/test/fz_http/gateways_test.exs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    test "create_gateway/1 with duplicate name returns an error changeset" do
+      _ = gateway(%{name: "gateway"})
+
+      duplicate_name = %{
+        name: "gateway",
+        registration_token: "test_token",
+        registration_token_created_at: DateTime.utc_now()
+      }
+
+      assert {:error, %Ecto.Changeset{}} = Gateways.create_gateway(duplicate_name)
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http/gateways_test.exs]: Explicit error check with `==`

File: apps/fz_http/test/support/fixtures/gateways_fixtures.ex [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  Test helpers for creating gateways with the `FzHttp.Gateways` context.
+  """
+
+  alias FzHttp.{Repo, Gateways, Gateways.Gateway}
+
+  @doc """
+  Generate a `Gateway`, using "default" name if not specified.
+  """
+  def gateway(attrs \\ %{}) do
+    name = attrs[:name] || "default"
```
Comment (jamilbk) [inline apps/fz_http/test/support/fixtures/gateways_fixtures.ex]: Redundant with the field default -- think you should remove the field default and force this to be passed in?

**Review by jamilbk — CHANGES_REQUESTED** [review]
A few comments

File: apps/fz_http/lib/fz_http/allow_rules/allow_rule.ex [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      :destination,
+      :port_range_start,
+      :port_range_end,
+      :protocol
+    ])
+    |> validate_required([:destination, :gateway_id])
+    |> check_constraint(
+      :allow_rules,
+      name: :port_range_with_optional_protocol,
+      message: "Port range start and end should be set or unset"
```
Comment (conectado) [inline apps/fz_http/lib/fz_http/allow_rules/allow_rule.ex]:
```suggestion
      message: "A port range needs both start and end. Additionally, a protocol requires a port range to be applied to."
```

**Review by conectado — APPROVED** [review]
LGTM.

Non-blocking nit picky suggestion left.

**Pull request merged**